bftengine/include/bftengine/Replica.hpp

```diff
@@ -75,6 +77,7 @@ class IReplica {
   // TODO(GG) : move the following methods to an "advanced interface"
   virtual void SetAggregator(std::shared_ptr<concordMetrics::Aggregator> a) = 0;
   virtual void restartForDebug(uint32_t delayMillis) = 0;  // for debug only.
+  virtual void setControlStateManager(std::shared_ptr<IControlStateManager> controlStateManager){};
```

Comment:
In general, `shared_ptr` makes ownership and lifetime of objects unclear. Do we really need it to be shared?
As far as I see, replica owns the manager and passes it to the command handlers.
If so, the manager may be moved to the replica and passed to the command handlers as reference(the lifetime of the handlers is supposed to be shorter than replicas so we're safe, I am not mistaken).
If so, there is no need to use pointers(only delete copy-constructors and operators of the manager base class).

Reply:
I was trying to use references instead of shared_ptrs but it looks like that using references complicates the code a lot.
Explanation:
The state transfer is initiated in the higher level of the kvbc replica, then we need to pass it through all bft layers and interfaces to the replicaBase and the implementations of the command handler.
If I use references I have to pass the controlStateManager through the constructors, which may complicate the current code flow, and also may depend on the order we initiate the objects (replica and command handler) in Concord (the product).
Thus, I prefer to use shared_ptr and getters and setters over references.

Reply:
I understand the complications but the question is if state transfer and command handler can exist w/o replace?
No pressure but I am really trying to understand it from functional perspective.